src/nodes.js

```diff
@@ -143,10 +143,55 @@ var registerChild = function(parent, key, child) {
 };
 
 
+/**
+ * Cleans a key map, nulling any nodes that are no longer in the DOM.
+ * @param {!Node} node
+ */
+var cleanKeyMap = function(node) {
+  var data = getData(node);
+  var keyMap = data.keyMap;
+  
+  if (!keyMap) {
+    return;
+  }
+
+  for (var key in keyMap) {
+    if (!keyMap[key].parentNode) {
+      delete keyMap[key];
+    }
+  }
+
+  data.keyMapValid = true;
+};
+
+
+/**
+ * Marks the key map for a given Element as invalid. It will be cleaned the
+ * next time it is needed.
+ * @param {!Node} node
+ */
+var invalidateKeyMap = function(node) {
+  getData(node).keyMapValid = false;
+};
+
+
+/**
+ * @param {!Node} node
+ * @return {boolean} True if the keyMap for the given Node is valid, false
+ *    otherwise.
+ */
+var isKeyMapValid = function(node) {
```

Comment:
> It feels a bit odd to directly access things related to the keyMap in some cases and not in others.

I think that's because we don't know upfront if the `keyMap` exists, so we lazily hide it behind `getKeyMap`. But, if we already know it does, I don't see the harm in mutating it directly (as long as it's in the `nodes` module). But, this is a `NodeData` property, not a `keyMap` one, so I don't see the harm in doing `data.keyMapIsValid`. Especially since the we've [already retrieved](https://github.com/google/incremental-dom/pull/101/files#diff-e46389c724a011cf4bc5a3bb0240b39cR126) `data`.

> I think I am going to change it to explicitly 'import' the patch subtree (like in #86)

The current implementation feels much cleaner to me. I [dislike](https://github.com/google/incremental-dom/pull/54) seeing all of those `__incrementalDOM*` dynamic property access.

> I am going to change it to explicitly 'import' the patch subtree (like in #86) and create the key maps up front in a later PR.

I think that'll negate the performance improvements of the this PR and #48, since it'll require an extra loop of `node.childNodes`.